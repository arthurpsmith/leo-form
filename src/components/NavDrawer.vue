<template>
  <v-navigation-drawer
    v-model="localDrawer"
    temporary
    app
  >
    <v-system-bar color="secondary"></v-system-bar>
<!--    <v-img :src="this.$logoImageUrl"></v-img>-->
    <v-list>
      <v-list-item class="py-6 py-xs-2">
        <app-logo></app-logo>
      </v-list-item>
      <template v-for="(item, index) in items">
        <v-list-item v-if="!item.items" :key="`item-${index}`" :href="item.url">
          <v-list-item-title>{{ item.label }}</v-list-item-title>
        </v-list-item>
        <template v-if="item.items">
          <v-list-group :key="`multi-level-group${index}`">
            <template v-slot:activator>
              <v-list-item-title>{{ item.label }}</v-list-item-title>
            </template>
            <v-list-item v-for="(multiLevelItem, index) in item.items" :key="`item-${index}`" class="pl-6" :href="multiLevelItem.url">
              <v-list-item-title>{{ multiLevelItem.label }}</v-list-item-title>
            </v-list-item>
          </v-list-group>
        </template>
      </template>
    </v-list>
    <v-divider></v-divider>
  </v-navigation-drawer>
</template>

<script lang="ts">
import { defineComponent } from '@vue/composition-api'
import AppLogo from '@/components/AppLogo.vue'

interface MenuItem {
  label: string, icon: string, url: string
}

interface MultiLevelMenuItem {
  label: string, items: MenuItem[]
}

interface MenuList {
  items: MenuItem[] | MultiLevelMenuItem[]
}
export default defineComponent({
  name: 'NavDrawer',
  components: { AppLogo },
  props: {
    value: {
      type: Boolean
    }
  },
  setup () {
    const items: Array<MenuItem | MultiLevelMenuItem> = []
    function createItem (index: number, text: string): MenuItem {
      const number: number = index + 1
      const label = `${text} ${number}`
      const icon = ''
      const url = `https://www.crossref.org/${number}`
      return { label, icon, url }
    }
    function createMultiLevelMenuItem (index: number, text: string, numberOfSubItems: number) : MultiLevelMenuItem {
      const number: number = index + 1
      const multiItems: MultiLevelMenuItem = {
        label: `${text} ${number}`,
        items: []
      }
      for (let j = 0; j < numberOfSubItems; j++) {
        multiItems.items.push(createItem(j, 'Multi level option'))
      }

      return multiItems
    }
    for (let i = 0; i < 6; i++) {
      if (i === 5) {
        const multiLevelMenuItem: MultiLevelMenuItem = createMultiLevelMenuItem(i, 'Option', 6)
        items.push(multiLevelMenuItem)
      }
      if (i !== 5) {
        items.push(createItem(i, 'Option'))
      }
    }
    return { items }
  },
  data: () => ({
    localDrawer: false
  }),
  watch: {
    value: function () {
      this.localDrawer = this.value
    },
    localDrawer: function () {
      this.$emit('input', this.localDrawer)
    }
  }
})
</script>

<style scoped>

</style>
