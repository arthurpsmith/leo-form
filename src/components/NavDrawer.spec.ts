import { createLocalVue, mount } from '@vue/test-utils'
import NavDrawer from '@/components/NavDrawer.vue'
import Vuetify from 'vuetify'
import CompositionApi from '@vue/composition-api'

describe('NavDrawer.vue', () => {
  const localVue = createLocalVue()
  localVue.use(CompositionApi)
  let vuetify: Vuetify
  function mountFunction (options = {}) {
    return mount(NavDrawer, {
      localVue,
      vuetify,
      ...options
    })
  }
  beforeEach(() => {
    vuetify = new Vuetify()
  })
  it('should work', () => {
    const wrapper = mountFunction()

    expect(wrapper.html()).toMatchSnapshot()
  })
  // Open and close on activator and outside component click better handled in an e2e test
  it('Expands and collapses menu item with children on click', async () => {
    const wrapper = mountFunction()
    const menuItemWithChildren = wrapper.find('.v-list-group')
    const menuItemElementActivator = menuItemWithChildren.find('.v-list-item')
    await menuItemElementActivator.trigger('click')
    expect(menuItemWithChildren.classes()).toContain('v-list-group--active')
    await menuItemElementActivator.trigger('click')
    expect(menuItemWithChildren.classes()).not.toContain('v-list-group--active')
  })
})
