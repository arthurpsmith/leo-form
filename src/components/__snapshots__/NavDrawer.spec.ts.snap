// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavDrawer.vue should work 1`] = `
<nav class="v-navigation-drawer v-navigation-drawer--close v-navigation-drawer--fixed v-navigation-drawer--is-mobile v-navigation-drawer--temporary theme--light" style="height: 100vh; top: 0px; transform: translateX(-100%); width: 256px;">
  <div class="v-navigation-drawer__content">
    <div class="v-system-bar theme--light secondary" style="height: 24px;"></div>
    <div class="v-list v-sheet theme--light">
      <div tabindex="-1" class="py-6 py-xs-2 v-list-item theme--light">
        <div>
          <div aria-label="Crossref logo" role="img" class="v-image v-responsive theme--light" style="min-height: 36px; min-width: 72px; max-height: 50px; max-width: 72px;">
            <div class="v-responsive__sizer" style="padding-bottom: 65.01950585175553%;"></div>
            <transition-stub name="fade-transition" mode="in-out">
              <div class="v-image__image v-image__image--preload v-image__image--cover" style="background-position: center center;"></div>
            </transition-stub>
            <div class="v-responsive__content"></div>
          </div>
        </div>
      </div> <a tabindex="0" href="https://www.crossref.org/1" class="v-list-item v-list-item--link theme--light">
        <div class="v-list-item__title">Option 1</div>
      </a>
      <!----><a tabindex="0" href="https://www.crossref.org/2" class="v-list-item v-list-item--link theme--light">
        <div class="v-list-item__title">Option 2</div>
      </a>
      <!----><a tabindex="0" href="https://www.crossref.org/3" class="v-list-item v-list-item--link theme--light">
        <div class="v-list-item__title">Option 3</div>
      </a>
      <!----><a tabindex="0" href="https://www.crossref.org/4" class="v-list-item v-list-item--link theme--light">
        <div class="v-list-item__title">Option 4</div>
      </a>
      <!----><a tabindex="0" href="https://www.crossref.org/5" class="v-list-item v-list-item--link theme--light">
        <div class="v-list-item__title">Option 5</div>
      </a>
      <!---->
      <!---->
      <div class="v-list-group">
        <div tabindex="0" aria-expanded="false" role="button" class="v-list-group__header v-list-item v-list-item--link theme--light">
          <div class="v-list-item__title">Option 6</div>
          <div class="v-list-item__icon v-list-group__header__append-icon"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-down theme--light"></i></div>
        </div>
        <transition-stub name="expand-transition">
          <!---->
        </transition-stub>
      </div>
    </div>
    <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
  </div>
  <div class="v-navigation-drawer__border"></div>
</nav>
`;
